document-store/src/main/java/org/hypertrace/core/documentstore/Collection.java

```diff
@@ -16,6 +17,19 @@
    */
   boolean upsert(Key key, Document document) throws IOException;
 
+  /**
+   * Upsert (create a new doc or update if one already exists) the given document into the doc store
+   * if condition is evaluated to true. Provides optimistic lock support for concurrency update.
+   *
+   * @param key Unique key of the document in the collection.
+   * @param document Document to be upserted.
+   * @param condition Filter condition to be evaluated, on success update the document
+   * @param isUpsert Optional parameter indicates insert or update, default is true
+   * @return True if success. False otherwise.
+   */
+  boolean upsert(Key key, Document document, Filter condition, @Nullable Boolean isUpsert)
```

Comment:
`@Nullable Boolean` sounds a bit weird. Can we just have `boolean isUpsert`? Or, there is a reason to maintain 3 states? `null`, `true` or `false`

Because, `null` and `true` has the same behavior here

Reply:
No. While adding this, I was still under discussion, should we expose this var at all or not? Removed nullable